### PR TITLE
MINOR: replace test "expected" parameter by assertThrows

### PR DESCRIPTION
**Pull request by chia7712**
`assertThrows` can make checks in testing more flexible and accurate so I think it is worth substituting `assertThrows` for "expected" parameter.

This PR includes following changes.

1. ```@Test(expected = Exception.class)``` is replaced by ```assertThrows```
1. remove reference to ```org.scalatest.Assertions``` 
1. change the magic code from 1 to 2 for ```testAppendAtInvalidOffset``` to test ZSTD
1. rename ```testMaybeAddPartitionToTransactionXXXX``` to ```testNotReadyForSendXXX```
1. increase maxBlockMs from 1s to 3s to avoid unexpected timeout from ```TransactionsTest#testTimeout``` 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/test/java/org/apache/kafka/clients/NetworkClientTest.java [hunk 1]
```diff
@@ -117,13 +118,12 @@ public void setup() {
...
     }
 
-    @Test(expected = IllegalStateException.class)
+    @Test
     public void testSendToUnreadyNode() {
         MetadataRequest.Builder builder = new MetadataRequest.Builder(Collections.singletonList("test"), true);
         long now = time.milliseconds();
         ClientRequest request = client.newClientRequest("5", builder, now, false);
-        client.send(request, now);
-        client.poll(1, time.milliseconds());
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/clients/NetworkClientTest.java:128]: We don't need this anymore?

Comment (chia7712) [reply clients/src/test/java/org/apache/kafka/clients/NetworkClientTest.java:128]: ```client.send(request, now);``` throws exception so ```client.poll(1, time.milliseconds());``` was never reached.

**Comment by chia7712** [comment]
@ijuma Any thoughts to this PR?

File: clients/src/test/java/org/apache/kafka/common/metrics/MetricsTest.java [hunk 2]
```diff
@@ -196,12 +197,12 @@ public void testHierarchicalSensors() {
...
     }
 
-    @Test(expected = IllegalArgumentException.class)
+    @Test
     public void testBadSensorHierarchy() {
         Sensor p = metrics.sensor("parent");
         Sensor c1 = metrics.sensor("child1", p);
         Sensor c2 = metrics.sensor("child2", p);
-        metrics.sensor("gc", c1, c2); // should fail
+        assertThrows(IllegalArgumentException.class, () -> metrics.sensor("gc", c1, c2)); // should fail
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/common/metrics/MetricsTest.java]: We can remove the comment, it's redundant.

File: clients/src/test/java/org/apache/kafka/common/record/ByteBufferLogInputStreamTest.java [hunk 3]
```diff
@@ -116,8 +117,7 @@ public void iteratorRaisesOnTooLargeRecords() {
         buffer.flip();
 
         ByteBufferLogInputStream logInputStream = new ByteBufferLogInputStream(buffer, 25);
-        assertNotNull(logInputStream.nextBatch());
-        logInputStream.nextBatch();
+        assertThrows(CorruptRecordException.class, logInputStream::nextBatch);
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/common/record/ByteBufferLogInputStreamTest.java:119]: Hmm, I think this was intended to fail on the second `nextBatch`, so maybe we should adjust the test.

File: core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala [hunk 4]
```diff
@@ -411,14 +411,13 @@ abstract class EndToEndAuthorizationTest extends IntegrationTestHarness with Sas
...
     */
-  @Test(expected = classOf[KafkaException])
+  @Test
   def testNoConsumeWithoutDescribeAclViaAssign(): Unit = {
     noConsumeWithoutDescribeAclSetup()
     val consumer = createConsumer()
     consumer.assign(List(tp).asJava)
     // the exception is expected when the consumer attempts to lookup offsets
-    consumeRecords(consumer)
-    confirmReauthenticationMetrics()
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala:421]: We should not remove this, right?

**Review by ijuma — COMMENTED** [review]
Thanks for the PR. Looks good overall, a few minor comments below.

One more thing: we seem to sometimes use the scalatest asserts and sometimes the JUnit ones for the Scala code. I would suggest using the JUnit ones consistently as it will make it easier to convert the tests to JUnit 5.

File: core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala (hunk 4 above)
Comment (chia7712) [reply core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala:421]: good catching!

**Comment by chia7712** [comment]
> One more thing: we seem to sometimes use the scalatest asserts and sometimes the JUnit ones for the Scala code. I would suggest using the JUnit ones consistently as it will make it easier to convert the tests to JUnit 5.

Except for ```org.scalatest.Assertions.withClue```, others package from ```org.scalatest.Assertions``` are replaced by Junit.  There are other tests using ```scalatest``` and I can remove ```scalatest``` from code base in follow-up.

**Comment by chia7712** [comment]
@ijuma the latest commit replaces all ExpectedException or @Test(expected = Exception.class) by assertThrows. It seems to me we can get closer to junit 5 if this PR is got into trunk. Furthermore, I'd like to file following PR to remove all junit 4 code.

1. Replace ```org.hamcrest``` by junit
1. Replace ScalaTest by junit
2. Replace junit 4 APIs by Junit 5 for each module
    1. rewrite parameter tests
    2. replace ```org.junit``` by ```org.junit.jupiter.api```
    3. replace ```IntegrationTest``` by tag ```Integration```
    4. replace ```useJUnitPlatform``` by ```useJUnit``` 
    5. remove ```testCompile libs.junitVintageEngine```

WDYT?

**Comment by ijuma** [comment]
@chia7712 Thanks. I was about halfway through the review of the second to last commit. :) Let's get this PR over the line and not change it further please. Yes, I had filed a few JIRAs for migrating to JUnit 5, see https://issues.apache.org/jira/browse/KAFKA-7339.

I think we should do it on a module per module basis to make it easier to review (the JIRAs are structured in that way). I was going to try the clients module after this PR is merged. Feel free to take one or more of the other ones. `core` would be a good candidate perhaps.

**Comment by chia7712** [comment]
> I was about halfway through the review of the second to last commit. :) Let's get this PR over the line and not change it further please. 

I am so sorry my force push obstructed you from reviewing. I will never do that again :(

**Comment by chia7712** [comment]
> Feel free to take one or more of the other ones. core would be a good candidate perhaps.

I’d like to take over both core module and stream module :)

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java [hunk 5]
```diff
@@ -192,11 +192,12 @@ public void testConstructorWithSerializers() {
         new KafkaProducer<>(producerProps, new ByteArraySerializer(), new ByteArraySerializer()).close();
     }
 
-    @Test(expected = ConfigException.class)
+    @Test
     public void testNoSerializerProvided() {
         Properties producerProps = new Properties();
         producerProps.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9000");
-        new KafkaProducer(producerProps);
+
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: Was this new line intentional?

File: clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java [hunk 6]
```diff
@@ -663,12 +623,6 @@ public void shouldPreserveOffsetsFromCommitByGroupMetadataOnAbortIfTransactionsA
         producer.beginTransaction();
 
         String group2 = "g2";
-        Map<TopicPartition, OffsetAndMetadata> groupCommit2 = new HashMap<TopicPartition, OffsetAndMetadata>() {
-            {
-                put(new TopicPartition(topic, 2), new OffsetAndMetadata(53L, null));
-                put(new TopicPartition(topic, 3), new OffsetAndMetadata(84L, null));
-            }
-        };
         producer.sendOffsetsToTransaction(groupCommit, new ConsumerGroupMetadata(group2));
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java]: I think this was intended to be `groupCommit2`, right?

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java [hunk 7]
```diff
@@ -444,34 +444,30 @@ public void testAddPartitionToTransactionRetainsRetryBackoffWhenPartitionsAlread
         assertEquals(DEFAULT_RETRY_BACKOFF_MS, handler.retryBackoffMs());
     }
 
-    @Test(expected = IllegalStateException.class)
+    @Test
     public void testMaybeAddPartitionToTransactionBeforeInitTransactions() {
-        transactionManager.failIfNotReadyForSend();
-        transactionManager.maybeAddPartitionToTransaction(new TopicPartition("foo", 0));
+        assertThrows(IllegalStateException.class, () -> transactionManager.failIfNotReadyForSend());
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java:449]: This change doesn't seem right. The test is intended to test `maybeAddPartitionToTransaction`.

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java [hunk 8]
```diff
@@ -444,34 +444,30 @@ public void testAddPartitionToTransactionRetainsRetryBackoffWhenPartitionsAlread
...
     }
 
-    @Test(expected = IllegalStateException.class)
+    @Test
     public void testMaybeAddPartitionToTransactionBeforeBeginTransaction() {
         doInitTransactions();
-        transactionManager.failIfNotReadyForSend();
-        transactionManager.maybeAddPartitionToTransaction(new TopicPartition("foo", 0));
+        assertThrows(IllegalStateException.class, () -> transactionManager.failIfNotReadyForSend());
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java:455]: Same as above.

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java [hunk 9]
```diff
@@ -444,34 +444,30 @@ public void testAddPartitionToTransactionRetainsRetryBackoffWhenPartitionsAlread
...
 
-    @Test(expected = KafkaException.class)
+    @Test
     public void testMaybeAddPartitionToTransactionAfterAbortableError() {
         doInitTransactions();
         transactionManager.beginTransaction();
         transactionManager.transitionToAbortableError(new KafkaException());
-        transactionManager.failIfNotReadyForSend();
-        transactionManager.maybeAddPartitionToTransaction(new TopicPartition("foo", 0));
+        assertThrows(KafkaException.class, () -> transactionManager.failIfNotReadyForSend());
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java:463]: Same as above.

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java [hunk 10]
```diff
@@ -444,34 +444,30 @@ public void testAddPartitionToTransactionRetainsRetryBackoffWhenPartitionsAlread
...
     }
 
-    @Test(expected = KafkaException.class)
+    @Test
     public void testMaybeAddPartitionToTransactionAfterFatalError() {
         doInitTransactions();
         transactionManager.transitionToFatalError(new KafkaException());
-        transactionManager.failIfNotReadyForSend();
-        transactionManager.maybeAddPartitionToTransaction(new TopicPartition("foo", 0));
+        assertThrows(KafkaException.class, () -> transactionManager.failIfNotReadyForSend());
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java:470]: Same as above.

**Review by ijuma — COMMENTED** [review]
I have reviewed about 60 files and left a few comments. It would be good to clearly mention the cases where the translation wasn't mechanism as they are the most likely sources of issues.

File: clients/src/test/java/org/apache/kafka/common/memory/GarbageCollectedMemoryPoolTest.java [hunk 11]
```diff
@@ -23,68 +23,74 @@
...
     }
 
-    @Test(expected = IllegalArgumentException.class)
-    public void testReleaseForeignBuffer() throws Exception {
+    @Test
+    public void testReleaseForeignBuffer() {
         GarbageCollectedMemoryPool pool = new GarbageCollectedMemoryPool(1000, 10, true, null);
         ByteBuffer fellOffATruck = ByteBuffer.allocate(1);
-        pool.release(fellOffATruck);
-        pool.close();
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/common/memory/GarbageCollectedMemoryPoolTest.java:83]: Do we still need to call `close` to avoid a leak?

File: clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsBuilderTest.java [hunk 12]
```diff
@@ -504,20 +510,21 @@ public void writePastLimit() {
...
 
-    @Test(expected = IllegalArgumentException.class)
+    @Test
     public void testAppendAtInvalidOffset() {
         ByteBuffer buffer = ByteBuffer.allocate(1024);
         buffer.position(bufferOffset);
 
         long logAppendTime = System.currentTimeMillis();
-        MemoryRecordsBuilder builder = new MemoryRecordsBuilder(buffer, RecordBatch.MAGIC_VALUE_V1, compressionType,
+        MemoryRecordsBuilder builder = new MemoryRecordsBuilder(buffer, RecordBatch.MAGIC_VALUE_V2, compressionType,
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsBuilderTest.java:519]: Why did we change the magic value version here?

File: clients/src/test/java/org/apache/kafka/common/security/authenticator/SaslServerAuthenticatorTest.java [hunk 13]
```diff
@@ -67,8 +68,7 @@ public void testOversizeRequest() throws IOException {
             invocation.<ByteBuffer>getArgument(0).putInt(SaslServerAuthenticator.MAX_RECEIVE_SIZE + 1);
             return 4;
         });
-        authenticator.authenticate();
-        verify(transportLayer).read(any(ByteBuffer.class));
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/common/security/authenticator/SaslServerAuthenticatorTest.java:71]: Why don't we need the `verify` call?

File: connect/api/src/test/java/org/apache/kafka/connect/data/SchemaBuilderTest.java [hunk 14]
```diff
@@ -293,16 +295,13 @@ public void testEmptyStruct() {
...
-                .field("id", SchemaBuilder.string().doc("").build())
-                .build();
-        final Struct struct = new Struct(schema)
-                .put("id", "testing");
-        struct.validate();
+        assertThrows(SchemaBuilderException.class, () -> SchemaBuilder.struct()
+            .name("testing")
+            .field("id", SchemaBuilder.string().doc("").build())
+            .field("id", SchemaBuilder.string().doc("").build())
+            .build());
```
Comment (ijuma) [inline connect/api/src/test/java/org/apache/kafka/connect/data/SchemaBuilderTest.java:304]: It seems that we are not testing the `validate` method anymore.

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java (hunk 10 above)
Comment (chia7712) [reply clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java:470]: https://github.com/apache/kafka/commit/717c55be971df862c55f55d245b9997f1d6f998c moved the check of state out of ```maybeAddPartitionToTransaction``` and it added ```failIfNotReadyForSend``` to all test cases to make them throw exception. I will update the test cases name to avoid confusion.

File: clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsBuilderTest.java (hunk 12 above)
Comment (chia7712) [reply clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsBuilderTest.java:519]: magic 1 does not support ```ZStandard```

File: connect/api/src/test/java/org/apache/kafka/connect/data/SchemaBuilderTest.java (hunk 14 above)
Comment (chia7712) [reply connect/api/src/test/java/org/apache/kafka/connect/data/SchemaBuilderTest.java:304]: ```StructTest``` has test cases for ```validate``` method.

**Comment by chia7712** [comment]
@ijuma Thanks for your great reviewing. I have addressed most review comment in latest commit.

> It would be good to clearly mention the cases where the translation wasn't mechanism as they are the most likely sources of issues.

Pardon me. I failed to catch your point.

File: clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsBuilderTest.java (hunk 12 above)
Comment (ijuma) [reply clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsBuilderTest.java:519]: Unrelated to the changes in this PR then, just something you noticed? Worth a line in the PR description to avoid confusion.

Comment (chia7712) [reply clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsBuilderTest.java:519]: will copy that

**Comment by chia7712** [comment]
rebase code to fix conflicts

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java [hunk 15]
```diff
@@ -49,18 +49,17 @@
...
+        assertThrows(StreamsException.class, () -> node.init(null));
     }
 
     @SuppressWarnings("unchecked")
-    @Test(expected = StreamsException.class)
+    @Test
     public void shouldThrowStreamsExceptionIfExceptionCaughtDuringClose() {
         final ProcessorNode node = new ProcessorNode("name", new ExceptionalProcessor(), Collections.emptySet());
-        node.init(null);
-        node.close();
```
Comment (ijuma) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java:63]: Should we keep the `close`?

File: core/src/test/scala/unit/kafka/zookeeper/ZooKeeperClientTest.scala [hunk 16]
```diff
@@ -715,7 +711,7 @@ class ZooKeeperClientTest extends ZooKeeperTestHarness {
 
   private def cleanMetricsRegistry(): Unit = {
     val metrics = KafkaYammerMetrics.defaultRegistry
-    metrics.allMetrics.keySet.forEach(metrics.removeMetric)
+    metrics.allMetrics.keySet.forEach(m => metrics.removeMetric(m))
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/zookeeper/ZooKeeperClientTest.scala]: Why did we change this?

File: core/src/test/scala/integration/kafka/api/TransactionsTest.scala [hunk 17]
```diff
@@ -407,46 +406,40 @@ class TransactionsTest extends KafkaServerTestHarness {
...
   def testAbortTransactionTimeout(): Unit = {
     testTimeout(true, producer => producer.abortTransaction())
   }
 
-  def testTimeout(needInitAndSendMsg: Boolean,
+  private def testTimeout(needInitAndSendMsg: Boolean,
                   timeoutProcess: KafkaProducer[Array[Byte], Array[Byte]] => Unit): Unit = {
-    val producer = createTransactionalProducer("transactionProducer", maxBlockMs =  1000)
-
+    val producer = createTransactionalProducer("transactionProducer", maxBlockMs = 3000)
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/TransactionsTest.scala:432]: Why did we increase maxBlockMs here?

File: core/src/test/scala/unit/kafka/utils/QuotaUtilsTest.scala [hunk 18]
```diff
@@ -17,24 +17,21 @@
...
+import java.util.concurrent.TimeUnit
 
 class QuotaUtilsTest {
 
   private val time = new MockTime
   private val numSamples = 10
   private val sampleWindowSec = 1
   private val maxThrottleTimeMs = 500
-  private val metricName = new MetricName("test-metric", "groupA", "testA", Map.empty.asJava)
+  private val metricName = new MetricName("test-metric", "groupA", "testA", java.util.Collections.emptyMap())
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/utils/QuotaUtilsTest.scala]: Why did you change this?

**Review by ijuma — COMMENTED** [review]
Reviewed a bit over 200 files and a few questions/comments.

File: core/src/test/scala/unit/kafka/zookeeper/ZooKeeperClientTest.scala (hunk 16 above)
Comment (chia7712) [reply core/src/test/scala/unit/kafka/zookeeper/ZooKeeperClientTest.scala]: IntelliJ can't be aware that lambda function :(

Will revert it.

File: core/src/test/scala/unit/kafka/server/DynamicBrokerConfigTest.scala [hunk 19]
```diff
@@ -343,7 +340,7 @@ class DynamicBrokerConfigTest {
 
     class TestAuthorizer extends Authorizer with Reconfigurable {
       @volatile var superUsers = ""
-      override def start(serverInfo: AuthorizerServerInfo): util.Map[Endpoint, _ <: CompletionStage[Void]] = Map.empty.asJava
+      override def start(serverInfo: AuthorizerServerInfo): util.Map[Endpoint, _ <: CompletionStage[Void]] = util.Collections.emptyMap()
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/server/DynamicBrokerConfigTest.scala]: Why did we change this?

File: core/src/test/scala/integration/kafka/api/TransactionsTest.scala (hunk 17 above)
Comment (chia7712) [reply core/src/test/scala/integration/kafka/api/TransactionsTest.scala:432]: ```1000``` is too small to complete ```createTransactionalProducer``` and it causes ```TimeoutException```. I don't think those test cases tried to test timeout on ```createTransactionalProducer```.

File: core/src/test/scala/unit/kafka/utils/QuotaUtilsTest.scala (hunk 18 above)
Comment (chia7712) [reply core/src/test/scala/unit/kafka/utils/QuotaUtilsTest.scala]: my bad and will revert it. (IntelliJ complained about the erased types)

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java (hunk 15 above)
Comment (chia7712) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java:63]: It fails to init so the close is unnecessary.

Comment (ijuma) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java:63]: I don't think this is necessarily the case. The failure during init can happen after some resources have been allocated, right?

File: core/src/test/scala/unit/kafka/utils/QuotaUtilsTest.scala (hunk 18 above)
Comment (ijuma) [reply core/src/test/scala/unit/kafka/utils/QuotaUtilsTest.scala]: I would ignore IntelliJ issues if it compiles without warnings. I think there are a few more changes like this in the middle of thousands of lines of diff. Something to avoid in the future. :)

Comment (chia7712) [reply core/src/test/scala/unit/kafka/utils/QuotaUtilsTest.scala]: Please give me one second to revert those changes :(

**Comment by chia7712** [comment]
I have reverted most unrelated changes.  see https://github.com/apache/kafka/pull/9520/commits/d0e2236e96db0b514f4d2c07641eb236feb1de74

**Comment by ijuma** [comment]
@chia7712 Can you please remove the last commit? It seems to have a mix of reverts and it's triggering changes to files I had already reviewed.

**Comment by chia7712** [comment]
> Can you please remove the last commit?

done!

**Review by ijuma — APPROVED** [review]
Thanks for the PR, LGTM. Will you do the `core` conversion to JUnit 5 next? We are looking into how to convert some of the tests to work in both raft and zk modes and it would be easier to do that with JUnit 5.

**Comment by ijuma** [comment]
Looking at the test results, it looks like there are some real failures. For example:

`Build / JDK 8 / kafka.utils.QuotaUtilsTest.testBoundedThrottleTimeThrowsExceptionIfProvidedNonRateMetric`

**Comment by chia7712** [comment]
> Will you do the core conversion to JUnit 5 next?

of course. Should we keep using ```ScalaTest``` and ```hamcrest```? They are not in conflict with Junit 5 but it results in various code style in testing.

**Comment by ijuma** [comment]
I would do the minimal set of changes to make the test suite pass with JUnit 5.

**Comment by chia7712** [comment]
fix conflicting files

**Comment by chia7712** [comment]
@ijuma Thanks for all your great reviews!
